src/components/tabs/ScatterPlotTab.tsx

```diff
@@ -411,78 +417,105 @@ export default memo(function ScatterPlotTab(props: ScatterPlotTabProps): ReactEl
+      colors = subsampleColorRamp(colorRamp, COLOR_RAMP_SUBSAMPLES);
+    }
 
-      // Sort data into buckets
-      for (let i = 0; i < xData.length; i++) {
-        const isOutOfRange = false; // TODO: Implement out of range filtering
-        const isOutlier = dataset.outliers ? dataset.outliers[objectIds[i]] : false;
-        const objectId = objectIds[i];
-
-        let bucketIndex;
-        if (isOutOfRange) {
-          bucketIndex = 0;
-        } else if (isOutlier) {
-          bucketIndex = 1;
-        } else {
-          bucketIndex = getBucketIndex(featureData.data[objectId], colorMinValue, colorMaxValue, colors.length) + 2;
-        }
+    const colorMinValue = isCategorical ? 0 : colorRampMin;
+    const colorMaxValue = isCategorical ? categories.length - 1 : colorRampMax;
 
-        const bucket = traceDataBuckets[bucketIndex];
-        bucket.x.push(xData[i]);
-        bucket.y.push(yData[i]);
-        bucket.objectIds.push(objectIds[i]);
-        bucket.trackIds.push(trackIds[i]);
+    // Make a bucket group for each ramp/palette color and for the out-of-range and outliers.
+    // index 0 = out of filter range, index 1 = outliers.
+    const traceDataBuckets: TraceData[] = [];
+    for (let i = 0; i < colors.length + 2; i++) {
+      let color: TraceData["color"];
+      let marker: TraceData["marker"];
+      if (i === 0) {
+        color = `#${viewerConfig.outOfRangeDrawSettings.color.getHexString()}`;
+        marker = { ...markerConfig, ...markerConfig.outOfRange };
+      } else if (i === 1) {
+        color = `#${viewerConfig.outlierDrawSettings.color.getHexString()}`;
+        marker = { ...markerConfig, ...markerConfig.outliers };
+      } else {
+        color = `#${colors[i - 2].getHexString()}`;
+        marker = markerConfig;
       }
+      if (usingOverrideColor) {
+        color = `#${overrideColor.getHexString()}`;
+      }
+      traceDataBuckets.push({ x: [], y: [], objectIds: [], trackIds: [], color, marker });
+    }
 
-      // Optionally delete the outlier and out of range buckets to hide the values.
-      if (viewerConfig.outlierDrawSettings.mode === DrawMode.HIDE) {
-        traceDataBuckets.splice(1, 1);
+    // Sort data into buckets
+    for (let i = 0; i < xData.length; i++) {
+      const objectId = objectIds[i];
+      const isOutlier = dataset.outliers ? dataset.outliers[objectId] : false;
+      const isOutOfRange = inRangeIds[objectId] === 0;
+
+      if (Number.isNaN(objectId) || objectId === undefined || objectId <= 0) {
+        continue;
       }
-      if (viewerConfig.outOfRangeDrawSettings.mode === DrawMode.HIDE) {
-        traceDataBuckets.splice(0, 1);
+
+      let bucketIndex;
+      if (isOutOfRange) {
+        bucketIndex = 0;
+      } else if (isOutlier) {
+        bucketIndex = 1;
+      } else if (usingOverrideColor) {
+        bucketIndex = 2;
+      } else {
+        bucketIndex = getBucketIndex(featureData.data[objectId], colorMinValue, colorMaxValue, colors.length) + 2;
```

Comment:
this +2 offset starts to look like a bookkeeping problem.  Maybe at least call it NUM_PRESET_COLOR_BUCKETS or something?

Reply:
Also could name 0=COLOR_INDEX_OUTOFRANGE and 1=COLOR_INDEX_OUTLIERS etc. to keep it consistent across a couple of places